include/dfrn.php

```diff
@@ -1045,4 +1067,1311 @@ function deliver($owner,$contact,$atom, $dissolve = false) {
+			/// - poco:urls
+			/// - poco:locality
+			/// - poco:region
+			/// - poco:country
+
+			// Save the keywords into the contact table
+			$tags = array();
+			$tagelements = $xpath->evaluate($element."/poco:tags/text()", $context);
+			foreach($tagelements AS $tag)
+				$tags[$tag->nodeValue] = $tag->nodeValue;
+
+			if (count($tags))
+				$contact["keywords"] = implode(", ", $tags);
+
+			// "dfrn:birthday" contains the birthday converted to UTC
+			$old_bdyear = $contact["bdyear"];
+
+			$birthday = $xpath->evaluate($element."/dfrn:birthday/text()", $context)->item(0)->nodeValue;
+
+			if (strtotime($birthday) > time()) {
+				$bd_timestamp = strtotime($birthday);
+
+				$contact["bdyear"] = date("Y", $bd_timestamp);
+			}
+
+			// "poco:birthday" is the birthday in the format "yyyy-mm-dd"
+			$value = $xpath->evaluate($element."/poco:birthday/text()", $context)->item(0)->nodeValue;
+
+			if (!in_array($value, array("", "0000-00-00"))) {
+				$bdyear = date("Y");
+				$value = str_replace("0000", $bdyear, $value);
+
+				if (strtotime($value) < time()) {
+					$value = str_replace($bdyear, $bdyear + 1, $value);
+					$bdyear = $bdyear + 1;
+				}
+
+				$contact["bd"] = $value;
+			}
+
+			if ($old_bdyear != $contact["bdyear"])
+				self::birthday_event($contact, $birthday);
+
+			// Get all field names
+			$fields = array();
+			foreach ($r[0] AS $field => $data)
+				$fields[$field] = $data;
+
+			unset($fields["id"]);
+			unset($fields["uid"]);
+			unset($fields["avatar-date"]);
+			unset($fields["name-date"]);
+			unset($fields["uri-date"]);
+
+			 // Update check for this field has to be done differently
+			$datefields = array("name-date", "uri-date");
+			foreach ($datefields AS $field)
+				if (strtotime($contact[$field]) > strtotime($r[0][$field]))
+					$update = true;
+
+			foreach ($fields AS $field => $data)
+				if ($contact[$field] != $r[0][$field]) {
+					logger("Difference for contact ".$contact["id"]." in field '".$field."'. Old value: '".$contact[$field]."', new value '".$r[0][$field]."'", LOGGER_DEBUG);
+					$update = true;
+				}
+
+			if ($update) {
+				logger("Update contact data for contact ".$contact["id"], LOGGER_DEBUG);
+
+				q("UPDATE `contact` SET `name` = '%s', `nick` = '%s', `about` = '%s', `location` = '%s',
+					`addr` = '%s', `keywords` = '%s', `bdyear` = '%s', `bd` = '%s',
+					`name-date`  = '%s', `uri-date` = '%s'
+					WHERE `id` = %d AND `network` = '%s'",
+					dbesc($contact["name"]), dbesc($contact["nick"]), dbesc($contact["about"]), dbesc($contact["location"]),
+					dbesc($contact["addr"]), dbesc($contact["keywords"]), dbesc($contact["bdyear"]),
+					dbesc($contact["bd"]), dbesc($contact["name-date"]), dbesc($contact["uri-date"]),
+					intval($contact["id"]), dbesc($contact["network"]));
+			}
+
+			update_contact_avatar($author["avatar"], $importer["uid"], $contact["id"],
+						(strtotime($contact["avatar-date"]) > strtotime($r[0]["avatar-date"])));
+
+			$contact["generation"] = 2;
```

Comment:
The generation is a sign for the reliability of the provided data. It is used in the socgraph.php to prevent that old contact data that was relayed over several servers can overwrite contact data that we received directly.

See here: https://github.com/friendica/friendica/blob/master/include/socgraph.php#L165-L170